### PR TITLE
Update for Emacs 29 where loaddefs-gen.el replaces autoloads.el

### PR DESCRIPTION
**Pull request by andersjohansson**
Fixes issue #985.

I believe this may be what is needed, but I haven’t tested it extensively yet.

### CODE REVIEW

File: straight.el [hunk 1]
```diff
@@ -5142,11 +5144,15 @@ modifies the build folder, not the original repository."
           ;; Actually generate the autoload file.
           ;; Emacs > 28.1 replaces `update-directory-autoloads' with
           ;; `make-directory-autoloads'
```
Comment (raxod502) [inline straight.el:5146]: Probably want to extend this comment.

File: straight.el [hunk 2]
```diff
@@ -5098,14 +5098,16 @@ modifies the build folder, not the original repository."
...
+  ;; time, and therefore `generated-autoload-file' is not defined as a
+  ;; variable. Thus Emacs generates bytecode corresponding to a
+  ;; lexical binding of `generated-autoload-file', and then chokes
+  ;; badly when `generated-autoload-file' turns into a dynamic
+  ;; variable at runtime.
   (eval-and-compile
-    (require 'autoload))
+    (condition-case nil
+        (require 'loaddefs-gen)
+      (error (require 'autoload))))
```
Comment (raxod502) [inline straight.el:5110]: Might be able to use the `NOERROR` parameter of `require` here.

**Review by raxod502 — APPROVED** [review]
:sparkles: Thanks!

File: straight.el (hunk 1 above)
Comment (andersjohansson) [reply straight.el]: Done!

File: straight.el (hunk 2 above)
Comment (andersjohansson) [reply straight.el:5110]: Ah, you’re right, I didn’t read the documentation for require well enough. That becomes a cleaner test.

**Comment by raxod502** [comment]
Thanks again.

**Pull request merged**